### ISSUE TITLE
Uploaded files with illegal characters silently fail leading to data loss

### ISSUE THREAD

**Issue by thjderjktyrjkt** [drakkan/sftpgo#1005]
### Description

If a Linux guest connects to SFTPGo on Windows, and uploads a file with "illegal" file name deemed by Windows, then it silently fails and the uploaded file is lost. An empty file with trimmed file name is created instead.

### Environment

**Server**

OS: Windows Server 2019
SFTPGo Version: 2.3.2-5f313cc

**Client**

OS: Linux
Software: rclone v1.59.1

### Steps to reproduce

**Server side:**

1. Setup SFTPGo server on Windows (Server edition shouldn't be necessary; Windows 10 is enough)
2. Configure the SFTPGo server so it serves a local folder on a NFTS file system over SFTP protocol

**Client side:**

3. Configure the client interactively with `rclone config`, create both a new sftp remote and a local remote
4. Create a new text file with the name `test:test.txt`, and put something inside

```bash
touch "test:test.txt"
cat > "test:test.txt" << EOF
TEST
1
EOF
```

5. Create another text file without `illegal` filename `testtest.txt`

```bash
touch "testtest.txt"
cat > "testtest.txt" << EOF
TEST
2
EOF
```

6. Upload both text files

```bash
rclone copy local:test:test.txt sftpgo:/
rclone copy local:testtest.txt sftpgo:/
```

**Server side:**

7. Check both files on server side

### Expected behavior

- Step 6. server side should intentionally fail the upload process for `test:test.txt`
- Only `testtest.txt` gets uploaded

### Current behavior

- Step 6 doesn't throw an error
- On server side, both `test` and `testtest.txt` gets created, but the former is an empty file

**Comment by thjderjktyrjkt** [comment]
Perhaps before accepting byte streams you need to try touching the file and see if target storage accepts it, and only starts writing to it if it's actually created. That way it'll eliminate any other potential filename limitation issues (other than NTFS).

**Comment by drakkan** [comment]
It looks like you are seeing [NTFS alternative files streams](https://github.com/golang/go/issues/13035). 

If from an sftp client you do something like this you should get uploaded content

```
sftp> get test:test.txt /tmp/a.txt
Fetching /test:test.txt to /tmp/a.txt
test:test.txt                                                                                  100%    2     1.4KB/s   00:00 
```

This seems something windows specific. I have the same behaviour if from a Windows command prompt I do something like this:

```
type nul > test:test.txt
```

**Comment by drakkan** [comment]
> Perhaps before accepting byte streams you need to try touching the file and see if target storage accepts it, and only starts writing to it if it's actually created. That way it'll eliminate any other potential filename limitation issues (other than NTFS).

NTFS seems to accept this name (in a strange way)

**Comment by thjderjktyrjkt** [comment]
Yeah that seems to be the issue. So it's indeed Windows-specific. Any chance we can see a workaround in SFTPGo? Even by just banning/replacing "illegal" characters on uploading?

**Comment by drakkan** [comment]
> Yeah that seems to be the issue. So it's indeed Windows-specific. Any chance we can see a workaround in SFTPGo? Even by just banning/replacing "illegal" characters on uploading?

No sorry, we should do several checks. If you try to upload a real illegal filename for example `test>` it will fail as expected.

You can add a file pattern restriction yourself for the affected users

**Comment by thjderjktyrjkt** [comment]
In case anyone else have the same issue, go to SFTPGo web admin panel, click `Users` and for every users, do the following:

- Expand `ACLs`
- Add a new file pattern

| directory path | pattern |        |        |
| -------------- | ------- | ------ | ------ |
| `/`            | `*:*`   | Denied | Hidden |

Then disconnect and reconnect clients. Now files with `:` character won't get uploaded.
